# Ejercicio 2

Ana es la gerente de una fábrica y quiere minimizar los costes de mano de obra y, a la vez, asegurar que en cada uno de los turnos ($m$ turnos) hay $n$ trabajadores para satisfacer la demanda de producción.

Disponemos de los siguientes datos:

$c_i$ ≡ coste de mano de obra del trabajador $i$.

$t_j$ ≡ demanda de personal en el turno $j$.

$d_{ij}$ ≡ disponibilidad del trabajador $i$ en el turno $j$.


## Formulación general

Sean las variables:

$b_{ij}$: el trabajador $i$ trabaja en el turno $j$

\begin{align*}
\underset{b_{ij}}{\min} & \quad \sum_{i=1}^{n}\sum_{j=1}^{m}c_{i}b_{ij}\\
\text{s.t.:}&\\
  &\sum_{i=1}^{n} b_{ij}d_{ij} = t_{j} \quad \forall j 1..m\\
  &x_{ij} \geq 0 \quad \forall ij\\
  &x_{ij} \leq 1 \quad \forall ij
\end{align*}

donde $n$ es la cantidad de trabajadores y $m$, la cantidad de turnos. 

## Implementación

En esta sección se da solución al problema utilizando la biblioteca _pulp_

In [1]:
from pulp import *

Se introducen los datos tomados del fichero "fabrica.dat"

In [2]:
n = 8 #i trabajadores
m = 4 #j turnos

workers = ['David', 'Enrique', 'Antonio', 'Lucia', 'Eva', 'Tomas', 'Juan', 'Ana']
shifts = [i for i in range(0,m)]

t = { 0: 4,
     1: 1,
     2: 5,
     3: 2 
}

d = { 
    'David': {0:1, 1:1, 2:0, 3:1},
    'Enrique': {0:0, 1:1, 2:1, 3:1}, 
    'Antonio': {0:0, 1:0, 2:1, 3:1},
    'Lucia': {0:0, 1:0, 2:1, 3:1},
    'Eva': {0:1, 1:0, 2:1, 3:1},
    'Tomas': {0:0, 1:1, 2:1, 3:1},
    'Juan': {0:1, 1:0, 2:1, 3:1},
    'Ana': {0:1, 1:1, 2:1, 3:0}
}

c = {
    'David': 20, 
    'Enrique': 15,
    'Antonio': 35,
    'Lucia': 35,
    'Eva': 25,
    'Tomas': 30,
    'Juan': 20,
    'Ana': 20   
}

Se define el problema y sus variables

In [3]:
prob = LpProblem('factory_problem', LpMinimize)

In [4]:
b = LpVariable.dicts("b",(workers,shifts),0,1, LpInteger)
b

{'David': {0: b_David_0, 1: b_David_1, 2: b_David_2, 3: b_David_3},
 'Enrique': {0: b_Enrique_0, 1: b_Enrique_1, 2: b_Enrique_2, 3: b_Enrique_3},
 'Antonio': {0: b_Antonio_0, 1: b_Antonio_1, 2: b_Antonio_2, 3: b_Antonio_3},
 'Lucia': {0: b_Lucia_0, 1: b_Lucia_1, 2: b_Lucia_2, 3: b_Lucia_3},
 'Eva': {0: b_Eva_0, 1: b_Eva_1, 2: b_Eva_2, 3: b_Eva_3},
 'Tomas': {0: b_Tomas_0, 1: b_Tomas_1, 2: b_Tomas_2, 3: b_Tomas_3},
 'Juan': {0: b_Juan_0, 1: b_Juan_1, 2: b_Juan_2, 3: b_Juan_3},
 'Ana': {0: b_Ana_0, 1: b_Ana_1, 2: b_Ana_2, 3: b_Ana_3}}

In [5]:
prob += lpSum([c[i]*b[i][j] for i in workers for j in shifts])
prob

factory_problem:
MINIMIZE
20*b_Ana_0 + 20*b_Ana_1 + 20*b_Ana_2 + 20*b_Ana_3 + 35*b_Antonio_0 + 35*b_Antonio_1 + 35*b_Antonio_2 + 35*b_Antonio_3 + 20*b_David_0 + 20*b_David_1 + 20*b_David_2 + 20*b_David_3 + 15*b_Enrique_0 + 15*b_Enrique_1 + 15*b_Enrique_2 + 15*b_Enrique_3 + 25*b_Eva_0 + 25*b_Eva_1 + 25*b_Eva_2 + 25*b_Eva_3 + 20*b_Juan_0 + 20*b_Juan_1 + 20*b_Juan_2 + 20*b_Juan_3 + 35*b_Lucia_0 + 35*b_Lucia_1 + 35*b_Lucia_2 + 35*b_Lucia_3 + 30*b_Tomas_0 + 30*b_Tomas_1 + 30*b_Tomas_2 + 30*b_Tomas_3 + 0
VARIABLES
0 <= b_Ana_0 <= 1 Integer
0 <= b_Ana_1 <= 1 Integer
0 <= b_Ana_2 <= 1 Integer
0 <= b_Ana_3 <= 1 Integer
0 <= b_Antonio_0 <= 1 Integer
0 <= b_Antonio_1 <= 1 Integer
0 <= b_Antonio_2 <= 1 Integer
0 <= b_Antonio_3 <= 1 Integer
0 <= b_David_0 <= 1 Integer
0 <= b_David_1 <= 1 Integer
0 <= b_David_2 <= 1 Integer
0 <= b_David_3 <= 1 Integer
0 <= b_Enrique_0 <= 1 Integer
0 <= b_Enrique_1 <= 1 Integer
0 <= b_Enrique_2 <= 1 Integer
0 <= b_Enrique_3 <= 1 Integer
0 <= b_Eva_0 <= 1 Integer
0 <=

Se añaden las restricciones

In [6]:
for j in shifts:
    prob += lpSum([d[i][j]*b[i][j] for i in workers]) == t[j]
prob

factory_problem:
MINIMIZE
20*b_Ana_0 + 20*b_Ana_1 + 20*b_Ana_2 + 20*b_Ana_3 + 35*b_Antonio_0 + 35*b_Antonio_1 + 35*b_Antonio_2 + 35*b_Antonio_3 + 20*b_David_0 + 20*b_David_1 + 20*b_David_2 + 20*b_David_3 + 15*b_Enrique_0 + 15*b_Enrique_1 + 15*b_Enrique_2 + 15*b_Enrique_3 + 25*b_Eva_0 + 25*b_Eva_1 + 25*b_Eva_2 + 25*b_Eva_3 + 20*b_Juan_0 + 20*b_Juan_1 + 20*b_Juan_2 + 20*b_Juan_3 + 35*b_Lucia_0 + 35*b_Lucia_1 + 35*b_Lucia_2 + 35*b_Lucia_3 + 30*b_Tomas_0 + 30*b_Tomas_1 + 30*b_Tomas_2 + 30*b_Tomas_3 + 0
SUBJECT TO
_C1: b_Ana_0 + b_David_0 + b_Eva_0 + b_Juan_0 = 4

_C2: b_Ana_1 + b_David_1 + b_Enrique_1 + b_Tomas_1 = 1

_C3: b_Ana_2 + b_Antonio_2 + b_Enrique_2 + b_Eva_2 + b_Juan_2 + b_Lucia_2
 + b_Tomas_2 = 5

_C4: b_Antonio_3 + b_David_3 + b_Enrique_3 + b_Eva_3 + b_Juan_3 + b_Lucia_3
 + b_Tomas_3 = 2

VARIABLES
0 <= b_Ana_0 <= 1 Integer
0 <= b_Ana_1 <= 1 Integer
0 <= b_Ana_2 <= 1 Integer
0 <= b_Ana_3 <= 1 Integer
0 <= b_Antonio_0 <= 1 Integer
0 <= b_Antonio_1 <= 1 Integer
0 <= b_Antonio_2 <

In [7]:
print(prob.solve())
print("Status:", LpStatus[prob.status])

print()
print('----Variables----')
for v in prob.variables():
    print(v.name, "=", v.varValue)

print()
print('----Solution----')
print(round(value(prob.objective),2))

1
Status: Optimal

----Variables----
b_Ana_0 = 1
b_Ana_1 = 0
b_Ana_2 = 1
b_Ana_3 = 0
b_Antonio_0 = 0
b_Antonio_1 = 0
b_Antonio_2 = 0
b_Antonio_3 = 0
b_David_0 = 1
b_David_1 = 0
b_David_2 = 0
b_David_3 = 1
b_Enrique_0 = 0
b_Enrique_1 = 1
b_Enrique_2 = 1
b_Enrique_3 = 1
b_Eva_0 = 1
b_Eva_1 = 0
b_Eva_2 = 1
b_Eva_3 = 0
b_Juan_0 = 1
b_Juan_1 = 0
b_Juan_2 = 1
b_Juan_3 = 0
b_Lucia_0 = 0
b_Lucia_1 = 0
b_Lucia_2 = 0
b_Lucia_3 = 0
b_Tomas_0 = 0
b_Tomas_1 = 0
b_Tomas_2 = 1
b_Tomas_3 = 0

----Solution----
245


El problema alcanza la solución óptima con un valor de 245 que se traduce en el coste mínimo con $n$ trabajadores en $m$ turnos teniendo en cuenta los costes, demanda y disponibilidad. A continuación, se escribe el resultado de manera más legible

In [8]:
sol = {v.name:v.varValue for v in prob.variables()}
for j in shifts:
    print(f"Turno {j}")
    for i in workers:
        bij = str(b[i][j])
        if sol[bij] == 1:
            print(f"{i} trabaja en el turno {j}")
    print()

Turno 0
David trabaja en el turno 0
Eva trabaja en el turno 0
Juan trabaja en el turno 0
Ana trabaja en el turno 0

Turno 1
Enrique trabaja en el turno 1

Turno 2
Enrique trabaja en el turno 2
Eva trabaja en el turno 2
Tomas trabaja en el turno 2
Juan trabaja en el turno 2
Ana trabaja en el turno 2

Turno 3
David trabaja en el turno 3
Enrique trabaja en el turno 3



## Nueva restricción

Se agrega la restricción:

#### Enrique no puede trabajar en turnos junto con David ni con Juan.

Definamos la relación simétrica **no_puede_trabajar_con**:

$no\_puede\_trabajar\_con(x, y)$ denota que el trabajador $x$ no puede trabajar con el trabajador $y$

La relación **no_puede_trabajar_con** es simétrica.

A continuación se listan los pares que pertenecen a la relación.

In [9]:
pairs = [('David', 'Enrique'), ('Enrique', 'Juan'), ('Enrique', 'David'), ('Juan', 'Enrique')]

def set_pairs(pairs, w1, w2):
    return 1 if (w1,w2) in pairs else 0

Se define la variable binaria:

\begin{align*}
p_{ik} = \left\{ \begin{array}{lcc}
             1 & si  (i,k) \in no\_puede\_trabajar\_con(i, k) \\
             0 & e.o.c
             \end{array}
   \right.
\end{align*}

donde $i$ y $k$ representan trabajadores.

La nueva restricción se puede escribir como:

\begin{align*}
\text{s.t.:}&\\
  &p_{ik} + b_{ij} + b_{ki} \leq 2 \quad \forall ijk
\end{align*}


## Implementación

A continuación se trasnforman los datos en un formato que faciliten su tratamiento. 

In [10]:
p = {w1:{w2: set_pairs(pairs, w1, w2) for w2 in workers if w1!=w2} for w1 in workers}

Se crea una copia del problema para no modificar el problema original al adicionar las nuevas restricciones.

In [11]:
new_prob = prob.copy()

In [12]:
workers_pairs = [ (i,k) for i in workers for k in workers if i!=k]
for j in shifts:
    for i,k in workers_pairs:
        new_prob += p[i][k] + b[i][j] + b[k][j] <= 2
new_prob

factory_problem:
MINIMIZE
20*b_Ana_0 + 20*b_Ana_1 + 20*b_Ana_2 + 20*b_Ana_3 + 35*b_Antonio_0 + 35*b_Antonio_1 + 35*b_Antonio_2 + 35*b_Antonio_3 + 20*b_David_0 + 20*b_David_1 + 20*b_David_2 + 20*b_David_3 + 15*b_Enrique_0 + 15*b_Enrique_1 + 15*b_Enrique_2 + 15*b_Enrique_3 + 25*b_Eva_0 + 25*b_Eva_1 + 25*b_Eva_2 + 25*b_Eva_3 + 20*b_Juan_0 + 20*b_Juan_1 + 20*b_Juan_2 + 20*b_Juan_3 + 35*b_Lucia_0 + 35*b_Lucia_1 + 35*b_Lucia_2 + 35*b_Lucia_3 + 30*b_Tomas_0 + 30*b_Tomas_1 + 30*b_Tomas_2 + 30*b_Tomas_3 + 0
SUBJECT TO
_C1: b_Ana_0 + b_David_0 + b_Eva_0 + b_Juan_0 = 4

_C2: b_Ana_1 + b_David_1 + b_Enrique_1 + b_Tomas_1 = 1

_C3: b_Ana_2 + b_Antonio_2 + b_Enrique_2 + b_Eva_2 + b_Juan_2 + b_Lucia_2
 + b_Tomas_2 = 5

_C4: b_Antonio_3 + b_David_3 + b_Enrique_3 + b_Eva_3 + b_Juan_3 + b_Lucia_3
 + b_Tomas_3 = 2

_C5: b_David_0 + b_Enrique_0 <= 1

_C6: b_Antonio_0 + b_David_0 <= 2

_C7: b_David_0 + b_Lucia_0 <= 2

_C8: b_David_0 + b_Eva_0 <= 2

_C9: b_David_0 + b_Tomas_0 <= 2

_C10: b_David_0 + b_Juan_

In [13]:
print(new_prob.solve())
print("Status:", LpStatus[new_prob.status])

print()
print('----Variables----')
for v in new_prob.variables():
    print(v.name, "=", v.varValue)

print()
print('----Solution----')
print(round(value(new_prob.objective),2))

1
Status: Optimal

----Variables----
b_Ana_0 = 1
b_Ana_1 = 0
b_Ana_2 = 1
b_Ana_3 = 0
b_Antonio_0 = 0
b_Antonio_1 = 0
b_Antonio_2 = 1
b_Antonio_3 = 0
b_David_0 = 1
b_David_1 = 0
b_David_2 = 0
b_David_3 = 0
b_Enrique_0 = 0
b_Enrique_1 = 1
b_Enrique_2 = 1
b_Enrique_3 = 1
b_Eva_0 = 1
b_Eva_1 = 0
b_Eva_2 = 1
b_Eva_3 = 1
b_Juan_0 = 1
b_Juan_1 = 0
b_Juan_2 = 0
b_Juan_3 = 0
b_Lucia_0 = 0
b_Lucia_1 = 0
b_Lucia_2 = 0
b_Lucia_3 = 0
b_Tomas_0 = 0
b_Tomas_1 = 0
b_Tomas_2 = 1
b_Tomas_3 = 0

----Solution----
265


In [14]:
sol = {v.name:v.varValue for v in new_prob.variables()}
for j in shifts:
    print(f"Turno {j}")
    for i in workers:
        bij = str(b[i][j])
        if sol[bij] == 1:
            print(f"{i} trabaja en el turno {j}")
    print()

Turno 0
David trabaja en el turno 0
Eva trabaja en el turno 0
Juan trabaja en el turno 0
Ana trabaja en el turno 0

Turno 1
Enrique trabaja en el turno 1

Turno 2
Enrique trabaja en el turno 2
Antonio trabaja en el turno 2
Eva trabaja en el turno 2
Tomas trabaja en el turno 2
Ana trabaja en el turno 2

Turno 3
Enrique trabaja en el turno 3
Eva trabaja en el turno 3



Se puede notar que el problema modificado también alcanza un valor óptimo, sin embargo, al añadir las nuevas restricciones el costo aumenta respecto al problema original. Asimismo, se puede observar una disposición diferente en los turnos en el problema modificado.